graph/iterator/map.go

```diff
@@ -0,0 +1,115 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Copyright 2009 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build !appengine,!safe
+
+package iterator
+
+import (
+	"unsafe"
+
+	"gonum.org/v1/gonum/graph"
+)
+
+// A mapIter is an iterator for ranging over a map.
+type mapIter struct {
+	m  *emptyInterface
+	it unsafe.Pointer
+}
+
+type emptyInterface struct {
+	typ, word unsafe.Pointer
+}
+
+// newMapIterNodes returns a range iterator for a map of nodes.
+func newMapIterNodes(m map[int64]graph.Node) *mapIter {
+	return &mapIter{m: eface(m)}
+}
+
+// newMapIterEdges returns a range iterator for a map of edges.
+// The returned mapIter must not have its node method called.
+func newMapIterEdges(m map[int64]graph.Edge) *mapIter {
+	return &mapIter{m: eface(m)}
+}
+
+// newMapIterWeightedEdges returns a range iterator for a map.
+// The returned mapIter must not have its node method called.
+func newMapIterWeightedEdges(m map[int64]graph.WeightedEdge) *mapIter {
+	return &mapIter{m: eface(m)}
+}
+
+// newMapIterLines returns a range iterator for a map of edges.
```

Comment:
A line is a kind of edge, so the comment is correct, but in graph theory, a line is an edge in a multigraph (the terminology is muddled and we needed to have something that differentiated between simple graph edges and multigraph edges). The other term that you might find is "arc" (edge in a directed graph - which we don't use because they behave the same way for us).

Reply:
Thanks.